### DISCUSSION TITLE
Does not seem to work on macOS

### DISCUSSION THREAD

**Discussion by basilfx** [muiriswoulfe/NuGet-Transitive-Dependency-Finder#142] [Q&A] [answered]
I'm looking into this project, and I don't seem to get it to work.

The only thing it prints, is `Commencing analysis...`, then it exits with code 0 (no error, no exception).

I therefore looked into the code, and observed the following:

* If I step through the code (especially `ProcessWrapper.WaitForExitAsync`), it 'works'. This makes me believe that this is a timing issue (e.g. the output is written and synced to disk after the process exited). I do see the output file appearing.
* If it completes, it then fails when trying to use the `ProcessWrapper` for a second time. The exception is: "Process is already associated with a real process, so the requested operation cannot be performed.".

Version info:

```
git rev-parse HEAD
61256a71fb2bafa5aa2e5421224c8475960a2df3
```

```
dotnet --list-sdks
2.2.402 [/usr/local/share/dotnet/sdk]
3.1.413 [/usr/local/share/dotnet/sdk]
5.0.207 [/usr/local/share/dotnet/sdk]
```

```
uname -a
Darwin MBP-van-Bas.lan 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64
```

**Comment by basilfx** [comment]
If I place some breakpoints in the `Dispose` methods, I do see that they are invoked before the next line after `ProcessWrapper.WaitForExitAsync` happens.

The second issue is resolved by ensuring a new `Process` instance is initiated when needed.

**Comment by basilfx** [comment]
Changing the Main method to as follows, solves the first issue. It seems that the service provider is already disposed while waiting for the process to finish.

```csharp
        public static int Main(string[] parameters) =>
            ProgramInitializer.Run(
                parameters,
                serviceProvider => ProgramInitializer.GetProgramRunner(serviceProvider).RunAsync().Wait());
```

**Comment by muiriswoulfe** [comment]
Hi @basilfx. Thanks for letting me know about the issue, and thanks for the additional debugging you performed.

I can confirm that there appears to be something wrong when I run it as well. Unfortunately, I haven't had much time to contribute to it lately and some error appears to have crept in with the updates to the dependencies.

I've worked through it now and I have a set of fixes in the [`muiriswoulfe/run-fixes`](https://github.com/muiriswoulfe/NuGet-Transitive-Dependency-Finder/tree/muiriswoulfe/run-fixes) branch. I'm not quite ready to commit those changes to `main` yet as I took a few shortcuts, e.g., deleting some unit tests, but I believe they should be sufficient for you to run the application.

Let me know how you get on.

**Reply by basilfx** [reply]
Thanks, I will look into this some time next week. I have also some proof-of-concept implementation to consider project references as well. That was actually my expectation of this tool ;-)

This means that if "Project A" includes "Project B" as a project reference, and both depend on "Package X" as a package reference, then "Package X" is marked as transitive for "Project A" now.

**Comment by muiriswoulfe** [comment, marked as answer]
The `main` branch now includes all the fixes.

**Reply by basilfx** [reply]
Thanks, I have checked it out, and it works!
